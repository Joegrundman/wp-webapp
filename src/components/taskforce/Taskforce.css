.header {
  color: #cdcdcd;
}

.subheaderWithArrows {
  display: flex;
  justify-content: space-between;
}

.subheaderCenterText {
  margin-top: 0;
  color: #cdcdcd;
}

.canvasContainer {
  border: 1px solid #cdcdcd;
  border-radius: 3px;
  position: relative;
  margin: 0 auto;
  overflow: scroll;
}

.taskforceCanvas {
  position: absolute;
  top: 0;
  left: 0;
}