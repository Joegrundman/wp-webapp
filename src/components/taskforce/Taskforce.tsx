import ArrowButton from 'Atoms/arrowbutton/ArrowButton';
import * as React from 'react'
import locals from './Taskforce.css'

interface ITaskforceState {
  currentCountry: number
}

const countries: string[]= [
  'Britain',
  'USA',
  'France',
  'Russia',
  'Germany',
  'Italy',
  'Japan'
]

class Taskforce extends React.Component<{}, ITaskforceState> {
  public tfBgContext: CanvasRenderingContext2D
  public tfContext: CanvasRenderingContext2D

  constructor(props: any) {
    super(props)
    this.state = {
      currentCountry: 0
    }
  }

  public componentDidMount (): void {
    const tfBgCanvas: HTMLCanvasElement = document.getElementById('taskforceCanvasBackground') as HTMLCanvasElement
    this.tfBgContext = tfBgCanvas.getContext('2d') as CanvasRenderingContext2D
    const tfCanvas: HTMLCanvasElement = document.getElementById('taskforceCanvas') as HTMLCanvasElement
    this.tfContext = tfCanvas.getContext('2d') as CanvasRenderingContext2D
    this.drawTaskforceBackground();
    // getGame().setSelectedShipyard(shipyards[this.state.currentShipyard], this.syContext)
  }

  public drawTaskforceBackground (): void {
    const tfLineHeight = 62.5
    const tfNumberCol = 32.5
    const numRows = 12

    this.tfBgContext.strokeStyle= '#cdcdcd'
    this.tfBgContext.lineWidth = 1

    this.tfBgContext.beginPath()
    this.tfBgContext.moveTo(tfNumberCol, 0)
    this.tfBgContext.lineTo(tfNumberCol, numRows * tfLineHeight)
    this.tfBgContext.closePath()
    this.tfBgContext.stroke()

    for (let i: number = 1; i < numRows; i++) {
      
      this.tfBgContext.beginPath()
      this.tfBgContext.moveTo(0, i * tfLineHeight)
      this.tfBgContext.lineTo(500, i * tfLineHeight)
      this.tfBgContext.closePath()
      this.tfBgContext.stroke()
    }

    const taskforces: string[] = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12']
    this.tfBgContext.font = "18px sans-serif"
    this.tfBgContext.fillStyle = '#cdcdcd'
    taskforces.forEach((tf: string, i: number): void => {
      this.tfBgContext.fillText(tf, i < 9 ? 10 : 5, 40.5 + (tfLineHeight * i))
    })
  }

  public changeCountryBack = () => {
    const currentCountry = this.state.currentCountry === 0 ?
    countries.length -1 :
    this.state.currentCountry - 1
    this.setState({
      currentCountry
    })
  }

  public changeCountryForward = () => {
    const currentCountry = this.state.currentCountry < countries.length - 1 ?
      this.state.currentCountry + 1 :
      0

    this.setState({
      currentCountry
    })
  }

  public render () {

    const tfContainerSize = {
      height: 470,
      width: 300
    }

    const tfCanvasSize = {
      height: 12 * 62.5,
      width: 500
    }
    return (
      <div>
        <h1 className={locals.header} >Taskforces</h1>
        <div className={locals.subheaderWithArrows}>
          <ArrowButton direction="left" action={this.changeCountryBack}/>
          <h2 className={locals.subheaderCenterText}>{countries[this.state.currentCountry]}</h2>
          <ArrowButton direction="right" action={this.changeCountryForward}/>
        </div>
        <div className={locals.canvasContainer} style={{height: tfContainerSize.height, width: tfContainerSize.width}}>
          <canvas
            className={locals.taskforceCanvas}
            id="taskforceCanvasBackground"
            height={tfCanvasSize.height}
            width={tfCanvasSize.width}/>
          <canvas
            className={locals.taskforceCanvas}
            id="taskforceCanvas"
            height={tfCanvasSize.height}
            width={tfCanvasSize.width}/>
        </div>
      </div>
    )
  }
}

export default Taskforce
