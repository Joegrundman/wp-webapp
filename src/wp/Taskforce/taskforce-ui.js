// WP.Taskforce.UI = {

// 	handleMenuButton: function () {
// 		$("#taskforce").dialog('open');
// 		$("#taskforce").dialog("option", "width", 498);
// 		$("#taskforce").dialog("option", "resizable", false);
// 		taskforce.handleTaskforceSelected("Britain");
// 		taskforce.draw();
// 	}
// }

// WP.Taskforce.Mouse = {}

// WP.Taskforce.prototype.draw = function () {
// 	var tf = game.selectedTaskforce;
// 	var units = new Array();

// 	for (var i = 0; i < tf.taskforceUnits.length; i++) {
// 		var taskforceUnit = tf.taskforceUnits[i];
// 		var unit = game.getUnitForTaskforce(taskforceUnit.id, taskforceUnit.x, taskforceUnit.y);
// 		units.push(unit);
// 	}
// 	var holder = WP.UnitHolder.unitHolderBuilder(taskforceCtx, $("#tfDetails"));
// 	// var holder = WP.UnitHolder.Util.unitHolderBuilder(taskforceCtx, $("#tfDetails"));
// 	holder.units = units;
// 	holder.drawTaskforce();
// 	taskforce.unitHolder = holder;
// }
