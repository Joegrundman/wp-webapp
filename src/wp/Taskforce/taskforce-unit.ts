'use strict'

class TaskforceUnit {
    public id: number;
    public x: number;
    public y: number;
    constructor (id: number, x: number, y: number){
        this.id = id;
        this.x = x;
        this.y = y;
    }
}

export default TaskforceUnit;
