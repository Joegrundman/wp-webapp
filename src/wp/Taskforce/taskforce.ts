import Unit from 'Unit/unit';
import TaskforceUnit from './taskforce-unit';

class Taskforce {

    public id: number;
    public owner: string;
    public size: number;
    public taskforceUnits: TaskforceUnit[];
    public unitHolder: object | null;
    public currentSquareX: number | null;
    public currentSquareY: number | null;
    public dragging: boolean;

    constructor (id: number, owner: string, size: number) {
        this.id = id;
        this.owner = owner;
        this.size = size;
        this.taskforceUnits = new Array();
        this.unitHolder = null;
        this.currentSquareX = null;
        this.currentSquareY = null;
        this.dragging = false;       
    }
    
    public addTaskforceUnit (taskforceUnit: TaskforceUnit) {
        this.taskforceUnits.push(taskforceUnit)
    }
  
    public removeUnitFrom (taskforceUnit: TaskforceUnit) {
        this.taskforceUnits = this.taskforceUnits.filter((tfUnit: TaskforceUnit) => taskforceUnit !== tfUnit)
    }

    public updateTaskforceUnitAddress (unit: Unit) {
        const tfUnit: TaskforceUnit | undefined = this.taskforceUnits.find((tfu: TaskforceUnit)=> tfu.id === unit.id)
        if (tfUnit && unit.holderX && unit.holderY) {
            tfUnit.x = unit.holderX;
            tfUnit.y = unit.holderY;
        }   
    }

    // public draw () {

    //     const units: Unit[] = [];

    //     this.taskforceUnits.forEach((tfu: TaskforceUnit) => {
    //         const unit: Unit = game.getUnitForTaskforce(taskforceUnit.id, taskforceUnit.x, taskforceUnit.y)
    //         units.push(unit);
    //     });
    //     for (var i = 0; i < tf.taskforceUnits.length; i++) {
    //         var taskforceUnit = tf.taskforceUnits[i];
    //         var unit = game.getUnitForTaskforce(taskforceUnit.id, taskforceUnit.x, taskforceUnit.y);
    //         units.push(unit);
    //     }
    //     var holder = WP.UnitHolder.unitHolderBuilder(taskforceCtx, $("#tfDetails"));
    //     holder.units = units;
    //     holder.drawTaskforce();
    //     taskforce.unitHolder = holder;        
    // }
    /**
     * Upon selecting a taskforce owner, this function draws a new flag and updates details
     * @param {object} owner -  the current owner of the taskforce
     */
    // handleTaskforceSelected (owner) {
    //     var taskforce = game.getTaskforceFromOwner(owner);
    //     var cty = game.getCountryFromName(owner);
    //     game.setSelectedTaskforce(taskforce);
    //     $("#tfCountryFlag").attr("src", WP.Country.UI.getFlagUrl(cty));
    //     taskforce.draw();
    // }
    // /**
    //  * Upon clicking the mouse button, this function finds the unit in the taskforce display under the mouse cursor
    //  * @param {object} e -  the mouse event object
    //  */    
    // onMouseDown (e) {
    //     var point = getPoint('taskforceCanvas', e);
    //     var stack = taskforce.unitHolder.findStackFor(point.x, point.y);
    //     if (stack) {
    //         game.setSelectedUnit(stack.getTopUnit());
    //         taskforce.unitHolder.drawStack(stack);
    //         taskforce.dragging = true;
    //     }
    //     else {
    //         game.setSelectedUnit(null);
    //     }       
    // }

    // /**
    //  * Upon double clicking the mouse, this function rotates the units in the stack under the mouse cursor
    //  */ 
    
    // onDoubleClick () {
    //     var unit = game.selectedUnit;
    //     var stack = taskforce.unitHolder.findStackContaining(unit);
    //     if (!stack) { return; }
    //     game.setSelectedUnit(null);
    //     if (stack.units.length > 1) {
    //         stack.rotateUnits();
    //         taskforce.unitHolder.drawStack(stack);
    //     }        
    // }
    // /**
    //  * Upon moving the mouse, this function drags the currently selected unit
    //  * @param {object} e - the mouse event object
    //  */     
    // onMouseMove (e) {
    //     taskforce.setCurrentSquare(e);
    //     if (game.selectedUnit && taskforce.dragging == true) {
    //         var unit = game.selectedUnit;
    //         if ((taskforce.currentSquareX != unit.holderX) || (taskforce.currentSquareY != unit.holderY)) {
    //             taskforce.moveUnit(unit);
    //         }
    //     }        
    // }
    // /**
    //  * Upon letting go of the mouse button, this function stops dragging the currently selected unit
    //  */     
    // onMouseUp () {
    //     taskforce.dragging = false
    // }
    // /**
    //  * Upon moving the mouse, this function finds the taskforce square currently under the cursor
    //  * @param {object} e - the mouse event object
    //  */     
    // setCurrentSquare (e) {
    //     var point = getPoint('taskforceCanvas', e);
    //     this.currentSquareX = Math.floor((point.x - 4) / 58);
    //     this.currentSquareY = Math.floor((point.y - 7) / 58);
    //     if (this.currentSquareY < 0) this.currentSquareY = 0;
    //     if (this.currentSquareX < 0) this.currentSquareX = 0;       
    // }
    // /**
    //  *  Updates the unit address to match its location after being dragged
    //  * @param {object} unit - the unit that has been moved
    //  */     
    // moveUnit (unit) {
    //     var stack = taskforce.unitHolder.findStackContaining(unit);
    //     stack.removeUnit(unit);
    //     unit.holderX = taskforce.currentSquareX;
    //     unit.holderY = taskforce.currentSquareY;
    //     taskforce.updateTaskforceUnitAddress(unit);
    //     taskforce.unitHolder.drawTaskforce();        
    // }
}

export default Taskforce;