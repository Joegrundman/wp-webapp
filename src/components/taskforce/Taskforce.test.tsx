import { shallow, ShallowWrapper } from 'enzyme'
import * as React from 'react'
import Taskforce from './Taskforce'


jest.mock('Wp/Game', () => ({
  getGame: jest.fn(() => ({
    setSelectedTaskforce: jest.fn()
  }))
}))

jest.spyOn(document, 'getElementById').mockImplementation(() => {
  return {
    getBoundingClientRect: jest.fn(() => ({x: 20, y: 30})),
    getContext: jest.fn(() => ({
      beginPath: jest.fn(),
      closePath: jest.fn(),
      drawImage: jest.fn(),
      fillText: jest.fn(),
      lineTo: jest.fn(),
      moveTo: jest.fn(),
      stroke: jest.fn()
    }))
  }
})

describe('<Taskforce />', (): void => {
  describe('@render', (): void => {
    it('should render as expected with default props', () => {
      const component: ShallowWrapper = shallow(<Taskforce />)
      expect(component).toMatchSnapshot()
    })
  })

  describe('methods', (): void => {
    describe('changeTaskforceOwnerBack', (): void => {
      it('should move to the previous country', () => {
        const component: ShallowWrapper = shallow(<Taskforce />)
        const instance = component.instance() as Taskforce
        instance.setState({ currentTaskforceOwner: 3 })
        instance.changeTaskforceOwnerBack()
        expect(instance.state.currentTaskforceOwner).toBe(2)
      })

      it('should cycle the currentTaskforceOwner to the end of the list if at zero', () => {
        const component: ShallowWrapper = shallow(<Taskforce />)
        const instance = component.instance() as Taskforce
        instance.setState({ currentTaskforceOwner:  0})
        instance.changeTaskforceOwnerBack()
        expect(instance.state.currentTaskforceOwner).toBe(6)
      })
    })

    describe('changeTaskforceOwnerForward', (): void => {
      it('should increment the currentTaskforceOwner', () => {
        const component: ShallowWrapper = shallow(<Taskforce />)
        const instance = component.instance() as Taskforce
        instance.setState({ currentTaskforceOwner: 5 })
        instance.changeTaskforceOwnerForward()
        expect(instance.state.currentTaskforceOwner).toBe(6)
      })

      it('should cycle the currentTaskforceOwner to the start of the list if at the end', () => {
        const component: ShallowWrapper = shallow(<Taskforce />)
        const instance = component.instance() as Taskforce
        instance.setState({ currentTaskforceOwner:  6})
        instance.changeTaskforceOwnerForward()
        expect(instance.state.currentTaskforceOwner).toBe(0)
      })
    })
  })
})