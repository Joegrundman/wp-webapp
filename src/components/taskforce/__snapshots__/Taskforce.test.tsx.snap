// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Taskforce /> @render should render as expected with default props 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Taskforce />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1
          className={undefined}
        >
          Taskforces
        </h1>,
        <div
          className={undefined}
        >
          <ArrowButton
            action={[Function]}
            direction="left"
          />
          <h2
            className={undefined}
          >
            Britain
          </h2>
          <ArrowButton
            action={[Function]}
            direction="right"
          />
        </div>,
        <div
          className={undefined}
          style={
            Object {
              "height": 470,
              "width": 300,
            }
          }
        >
          <canvas
            className={undefined}
            height={750}
            id="taskforceCanvasBackground"
            width={500}
          />
          <canvas
            className={undefined}
            height={750}
            id="taskforceCanvas"
            width={500}
          />
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Taskforces",
          "className": undefined,
        },
        "ref": null,
        "rendered": "Taskforces",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <ArrowButton
              action={[Function]}
              direction="left"
            />,
            <h2
              className={undefined}
            >
              Britain
            </h2>,
            <ArrowButton
              action={[Function]}
              direction="right"
            />,
          ],
          "className": undefined,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "action": [Function],
              "direction": "left",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Britain",
              "className": undefined,
            },
            "ref": null,
            "rendered": "Britain",
            "type": "h2",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "action": [Function],
              "direction": "right",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <canvas
              className={undefined}
              height={750}
              id="taskforceCanvasBackground"
              width={500}
            />,
            <canvas
              className={undefined}
              height={750}
              id="taskforceCanvas"
              width={500}
            />,
          ],
          "className": undefined,
          "style": Object {
            "height": 470,
            "width": 300,
          },
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": undefined,
              "height": 750,
              "id": "taskforceCanvasBackground",
              "width": 500,
            },
            "ref": null,
            "rendered": null,
            "type": "canvas",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": undefined,
              "height": 750,
              "id": "taskforceCanvas",
              "width": 500,
            },
            "ref": null,
            "rendered": null,
            "type": "canvas",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1
            className={undefined}
          >
            Taskforces
          </h1>,
          <div
            className={undefined}
          >
            <ArrowButton
              action={[Function]}
              direction="left"
            />
            <h2
              className={undefined}
            >
              Britain
            </h2>
            <ArrowButton
              action={[Function]}
              direction="right"
            />
          </div>,
          <div
            className={undefined}
            style={
              Object {
                "height": 470,
                "width": 300,
              }
            }
          >
            <canvas
              className={undefined}
              height={750}
              id="taskforceCanvasBackground"
              width={500}
            />
            <canvas
              className={undefined}
              height={750}
              id="taskforceCanvas"
              width={500}
            />
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Taskforces",
            "className": undefined,
          },
          "ref": null,
          "rendered": "Taskforces",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <ArrowButton
                action={[Function]}
                direction="left"
              />,
              <h2
                className={undefined}
              >
                Britain
              </h2>,
              <ArrowButton
                action={[Function]}
                direction="right"
              />,
            ],
            "className": undefined,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "action": [Function],
                "direction": "left",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Britain",
                "className": undefined,
              },
              "ref": null,
              "rendered": "Britain",
              "type": "h2",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "action": [Function],
                "direction": "right",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <canvas
                className={undefined}
                height={750}
                id="taskforceCanvasBackground"
                width={500}
              />,
              <canvas
                className={undefined}
                height={750}
                id="taskforceCanvas"
                width={500}
              />,
            ],
            "className": undefined,
            "style": Object {
              "height": 470,
              "width": 300,
            },
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": undefined,
                "height": 750,
                "id": "taskforceCanvasBackground",
                "width": 500,
              },
              "ref": null,
              "rendered": null,
              "type": "canvas",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": undefined,
                "height": 750,
                "id": "taskforceCanvas",
                "width": 500,
              },
              "ref": null,
              "rendered": null,
              "type": "canvas",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
